### PR TITLE
Update the tf2 documentation

### PR DESCRIPTION
**Pull request by clalancette**
This PR is a logical continuation of #462 from @kurshakuz .

In particular, what it does is describe some high-level concepts that inform the design of tf2.

Where this PR differs significantly from #462 is that it is fully integrated into the rosdoc2 pipeline.  So once this is approved and merged, it will show up automatically on https://docs.ros.org/en/rolling/p/tf2/

### CODE REVIEW

File: tf2/mainpage.dox [hunk 1]
```diff
@@ -1,36 +1,195 @@
...
-conversion interface to apply transformations from the tf server to these custom datatypes.
+tf2 offers a templated conversion interface for external libraries to specify conversions between ``tf2``-specific data types and user-defined data types.
+Various templated functions in ``tf2_ros`` use the conversion interface to apply transformations from the tf server to these custom datatypes.
 
 The conversion interface is defined in tf2/convert.h.
 
-Some packages that implement this interface:
+\subsection buffer-core-relations Buffer Core: Record and lookup relations between frames
+
+The ``tf2`` library implements the interface defined by tf2::BufferCore.
```
Comment (ahcorde) [inline tf2/mainpage.dox]:
```suggestion
The ``tf2`` library implements the interface defined by `tf2::BufferCore`.
```

File: tf2/mainpage.dox [hunk 2]
```diff
@@ -1,36 +1,195 @@
...
+
+``tf2`` assumes that there is a tree of coordinate frame transforms which define the relationship between all coordinate frames.
+After transformation relationships are supplied, query specifiyng target frame, source frame, and time point can be used to obtain required data.
+``tf2`` will take care of all the intermediate transfromation steps for specific queries.
+
+Additionally, ``tf2`` features data interpolation.
+The probability that a specific query would be at the same timestamp of all the frames in the system is very unlikely in an asynchronous system with nanosecond precision.
+Therefore, for any given query it can be expected that data is interpolated.
+
+It should be noted that buffer implicitly limits the maximum cache size of 10s by default as defined by the tf2::TIMECACHE_DEFAULT_MAX_STORAGE_TIME and it cannot interpolate outside of the cache history.
```
Comment (ahcorde) [inline tf2/mainpage.dox]:
```suggestion
It should be noted that buffer implicitly limits the maximum cache size of 10s by default as defined by the `tf2::TIMECACHE_DEFAULT_MAX_STORAGE_TIME` and it cannot interpolate outside of the cache history.
```

File: tf2/mainpage.dox [hunk 3]
```diff
@@ -1,36 +1,195 @@
...
+Thus there is a risk of incurring extrapolation limits based on specific system.
+
+\subsubsection buffer-core-methods Buffer Core Methods
+
+The tf2::BufferCore class contains useful methods to update the existing tf buffer.
```
Comment (ahcorde) [inline tf2/mainpage.dox]:
```suggestion
The `tf2::BufferCore` class contains useful methods to update the existing tf buffer.
```

File: tf2/mainpage.dox [hunk 4]
```diff
@@ -1,36 +1,195 @@
...
+
+  - ``tf2`` methods to work with bullet datatypes natively in C++.
+
 - tf2_eigen
+
+  - ``tf2`` methods to work with Eigen datatypes natively in C++.
+
 - tf2_geometry_msgs
+
+  - ``tf2`` methods to work with geometry_msgs datatypes natively in C++ or Python.
```
Comment (ahcorde) [inline tf2/mainpage.dox]:
```suggestion
  - ``tf2`` methods to work with `geometry_msgs` datatypes natively in C++ or Python.
```

File: tf2/mainpage.dox (hunk 3 above)
Comment (ahcorde) [reply tf2/mainpage.dox]: and in the following methods too

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: tf2/mainpage.dox (hunk 1 above)
Comment (clalancette) [reply tf2/mainpage.dox]: Fixed (along with the others) in 23e606d

**Review by ahcorde — APPROVED** [review]

**Review by tfoote — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19897/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14381/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20604/)

**Pull request merged**